### ISSUE TITLE
How to reproduce the results of loading 70B LLMs within 300-700ms?

### ISSUE THREAD

**Issue by duanzhaol** [ServerlessLLM/ServerlessLLM#14]
**Describe the bug**
Hi, thanks for the interesting work. I'm trying to run your system following this [guide](https://serverlessllm.github.io/docs/stable/store/quickstart/). I want to quickly load the llama3-70B model on an 8*A100 server, but my loading time is more than 6.32 seconds instead of milliseconds. Is there any specific step I might be missing to achieve the results 300-700ms?

Specifically, the first time loading takes more than 20 seconds because it is loading from disk, and then it stabilizes at 5-6 seconds, likely because it is loading from CPU memory.

**To Reproduce**
I followed the [guide](https://serverlessllm.github.io/docs/stable/store/quickstart/) completely, only changing the model path.

**Expected behavior**
The model should load at the millisecond level.

**Screenshots**
![image](https://github.com/user-attachments/assets/08c5dfa9-308c-4b73-865f-1d269ee568e3)

**Server Info:**
 - OS: Ubuntu 22.04
- GPU:  8*A100 SXM 80GB server on Runpod

**Comment by future-xy** [comment]
Hi! Thank you for bringing this to our attention!

The loading performance can be influenced by several factors. To potentially improve this, you might want to consider the following suggestions:
1. Update the CUDA warmup section in your code as follows:
```python
num_gpus = torch.cuda.device_count()
for i in range(num_gpus):
    torch.ones(1).to(f"cuda:{i}")
    torch.cuda.synchronize()
```
This adjustment ensures that each GPU is properly warmed up, which is not covered in our current guide. We will update the documentation shortly to reflect this.

2. It's worth noting that some GPU servers share PCIe links among two or more GPUs, which can result in reduced bandwidth for each GPU. To help us better understand if this is affecting your performance, could you please provide a log or screenshot of the sllm-store-server?

Regarding the "300-700ms" latency, this was observed for loading cached model parameters (specifically for the opt-66b model) in our experimental prototype. For the current open-source version, the end-to-end model loading time on our servers is approximately 2-3 seconds for 70B models. We will make sure to update the README for clarity.

**Comment by duanzhaol** [comment]
Thank you for the detailed response. After modifying the warm-up code, the load time has significantly improved.

My machine configuration connects every two GPUs on a single PCIe link, and avoiding loading on GPUs sharing the same PCIe link has also effectively enhanced performance. Here are my test results:
| GPUs            | load time (s) |
|-----------------|--------------|
| 2-GPU           | 7.02         |
| 2-GPU-Diff-Link | 4.8          |
| 4-GPU           | 5.71         |
| 4-GPU-Diff-Link | 4.44         |
| 8-GPU           | 3.99         |

Additionally, I have another question: As far as I know, in common GPU servers, the main connection method is two GPUs connected on a single PCIe link. Is it a common type that each GPU have its own dedicated PCIe link? (I am also working on a cold start related project, and I am not sure if making such an assumption is reasonable. Looking forward to your guidance.)

**Comment by future-xy** [comment]
Hi! Thanks for such detailed measurements! This is valuable feedback.

In all our development and experiments, the GPU servers provide exclusive PCIe links for each GPU. However, it is possible to have multiple GPUs connected to the host on the same PCIe link. In such cases, some of our assumptions may break, leading to sub-optimal performance.

Would you be interested in helping to improve the ServerlessLLM store for these scenarios? We'd love to have you join the ServerlessLLM community and collaborate with us.
